powerauth-restful-security-javaee/src/main/java/io/getlime/security/powerauth/rest/api/jaxrs/service/v3/RecoveryService.java

```diff
@@ -65,19 +65,19 @@ public EciesEncryptedResponse confirmRecoveryCode(EciesEncryptedRequest request,
             final String applicationKey = httpHeader.getApplicationKey();
             if (activationId == null || applicationKey == null || request.getEphemeralPublicKey() == null
                     || request.getEncryptedData() == null || request.getMac() == null) {
-                logger.error("PowerAuth confirm recovery failed because of invalid request");
-                throw new PowerAuthAuthenticationException();
+                logger.warn("PowerAuth confirm recovery failed because of invalid request");
+                throw new PowerAuthAuthenticationException("POWER_AUTH_REQUEST_INVALID");
             }
             PowerAuthPortV3ServiceStub.ConfirmRecoveryCodeResponse paResponse = powerAuthClient.confirmRecoveryCode(activationId, applicationKey,
                     request.getEphemeralPublicKey(), request.getEncryptedData(), request.getMac(), request.getNonce());
             if (!paResponse.getActivationId().equals(activationId)) {
-                logger.error("PowerAuth confirm recovery failed because of invalid activation ID in response");
-                throw new PowerAuthAuthenticationException();
+                logger.warn("PowerAuth confirm recovery failed because of invalid activation ID in response");
+                throw new PowerAuthAuthenticationException("POWER_AUTH_REQUEST_INVALID");
             }
             return new EciesEncryptedResponse(paResponse.getEncryptedData(), paResponse.getMac());
         } catch (Exception ex) {
-            logger.warn("PowerAuth confirm recovery failed", ex);
-            throw new PowerAuthAuthenticationException();
+            logger.warn("PowerAuth confirm recovery failed, error: {}", ex.getMessage());
+            throw new PowerAuthAuthenticationException("POWER_AUTH_RECOVERY_ERROR");
```

Comment:
Let's pass `ex` as a parameter to the new exception to keep the cause somewhere.